level.go

```diff
@@ -108,20 +109,25 @@ func (l *LRU) Add(_ context.Context, k Key, e *Entry, ttl time.Duration) error {
 
 // Get gets an entry from the cache.
 func (l *LRU) Get(_ context.Context, k Key) (*Entry, error) {
-	l.lock.Lock()
-	defer l.lock.Unlock()
+	l.mu.RLock()
 	e, ok := l.Cache.Get(k)
+	l.mu.RUnlock()
+
 	if !ok {
 		return nil, ErrNotFound
 	}
+
 	switch e := e.(type) {
 	case *Entry:
 		return e, nil
 	case *entry:
 		if time.Now().Before(e.expiry) {
 			return e.Entry, nil
 		}
+
+		l.mu.Lock()
 		l.Cache.Remove(k)
+		l.mu.Unlock()
```

Comment:
Correct me if I am wrong, but the `Lock` here will wait for all read locks to be released before attempting to write, correct? If so, this is a deadlock, no?

This is the reason I used the `sync.Mutex` instead of a `sync.RWMutex` in the first place.

https://pkg.go.dev/sync#RWMutex.Lock